code/espurna/sensors/EZOPHSensor.h

```diff
@@ -0,0 +1,142 @@
+// EZO™ pH Circuit from Atlas Scientific
+//
+// Uses SoftwareSerial library
+// Copyright (C) 2018 by Rui Marinho <ruipmarinho at gmail dot com>
+// -----------------------------------------------------------------------------
+
+#if SENSOR_SUPPORT && EZOPH_SUPPORT
+
+#pragma once
+
+#include "Arduino.h"
+#include "BaseSensor.h"
+#include <SoftwareSerial.h>
+
+class EZOPHSensor : public BaseSensor {
+
+    public:
+
+        // ---------------------------------------------------------------------
+        // Public
+        // ---------------------------------------------------------------------
+
+        EZOPHSensor(): BaseSensor() {
+            _count = 1;
+            _sensor_id = SENSOR_EZOPH_ID;
+        }
+
+        ~EZOPHSensor() {
+            if (_serial) delete _serial;
+        }
+
+        // ---------------------------------------------------------------------
+
+        void setRX(unsigned char pin_rx) {
+            if (_pin_rx == pin_rx) return;
+            _pin_rx = pin_rx;
+            _dirty = true;
+        }
+
+        void setTX(unsigned char pin_tx) {
+            if (_pin_tx == pin_tx) return;
+            _pin_tx = pin_tx;
+            _dirty = true;
+        }
+
+        // ---------------------------------------------------------------------
+
+        unsigned char getRX() {
+            return _pin_rx;
+        }
+
+        unsigned char getTX() {
+            return _pin_tx;
+        }
+
+        // ---------------------------------------------------------------------
+        // Sensor API
+        // ---------------------------------------------------------------------
+
+        // Initialization method, must be idempotent
+        void begin() {
+            if (!_dirty) return;
+
+            if (_serial) delete _serial;
+
+            _serial = new SoftwareSerial(_pin_rx, _pin_tx);
+            _serial->begin(9600);
+
+            _ready = true;
+            _dirty = false;
+        }
+
+        // Descriptive name of the sensor
+        String description() {
+            char buffer[28];
+            snprintf(buffer, sizeof(buffer), "EZOPH @ SwSerial(%u,%u)", _pin_rx, _pin_tx);
+            return String(buffer);
+        }
+
+        // Descriptive name of the slot # index
+        String slot(unsigned char index) {
+            return description();
+        };
+
+        // Address of the sensor (it could be the GPIO or I2C address)
+        String address(unsigned char index) {
+            char buffer[6];
+            snprintf(buffer, sizeof(buffer), "%u:%u", _pin_rx, _pin_tx);
+            return String(buffer);
+        }
+
+        // Type for slot # index
+        unsigned char type(unsigned char index) {
+            if (index == 0) return MAGNITUDE_PH;
+            return MAGNITUDE_NONE;
+        }
+
+        void pre() {
+            _read();
+        }
+
+        // Current value for slot # index
+        double value(unsigned char index) {
+            if (index == 0) return _ph;
```

Comment:
@mcspr how many times is `_pre` called? It looked like it was only once during sensor initialization but the value on the UI is being updated. Maybe the `_read()` function should also be called inside `value()`?

Reply:
I will suggest passing `-DSENSOR_READ_INTERVAL=1` then.

Reply:
@mcspr even with `-DSENSOR_READ_INTERVAL=1 -DSENSOR_REPORT_EVERY=1`, `pre()` is only being called every 7-8 seconds. Any idea?

Reply:
Both defines are not modifiable after all. No one ever had to change these 🤔?

Reply:
Maybe `snsRead` is set from settings? Define is only the default value for it

Reply:
That's right. Is there anything else missing from this PR?